### PR TITLE
KOGITO-2024, KOGITO-2068, KOGITO-2110 - Add human task and rules (drl) support for Serverless Workflow

### PR DESCRIPTION
**Pull request by tsurdilo**
This also allows user tasks to be placed inside embedded subprocess (tested by the serverless workflow tests)

### CODE REVIEW

File: jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/util/ServerlessWorkflowUtils.java [hunk 1]
```diff
@@ -155,16 +156,22 @@ public static String scriptFunctionScript(String script) {
     }
 
     public static String conditionScript(String path, DefaultChoice.Operator operator, String value) {
+        String workflowVar = ServerlessWorkflowFactory.DEFAULT_WORKFLOW_VAR;
 
         if (path.startsWith("$.")) {
             path = path.substring(2);
+        } else {
```
Comment (cristianonicolai) [inline jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/util/ServerlessWorkflowUtils.java:163]:
```suggestion
        } else if {
```

File: jbpm/jbpm-serverless-workflow/src/test/java/org/jbpm/serverless/workflow/WorkflowFactoryTest.java [hunk 2]
```diff
@@ -232,4 +232,73 @@ public void testJoinNode() {
...
+        assertThat(process.getVariableScope().getVariables().size()).isEqualTo(1);
+    }
+
+    @Test
+    public void testHumanTaskNode() {
+        TestNodeContainer nodeContainer = new TestNodeContainer();
+        RuleFlowProcess process = new RuleFlowProcess();
+
+        Function function = new Function().withName("testfunction1").withMetadata(
+                new HashMap() {{
```
Comment (cristianonicolai) [inline jbpm/jbpm-serverless-workflow/src/test/java/org/jbpm/serverless/workflow/WorkflowFactoryTest.java:252]: My personal preference here would be to simply create a new HashMap and assign a variable, as this style endups creating an inner class.

File: jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/util/ServerlessWorkflowUtils.java (hunk 1 above)
Comment (tsurdilo) [reply jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/util/ServerlessWorkflowUtils.java]: updated

File: jbpm/jbpm-serverless-workflow/src/test/java/org/jbpm/serverless/workflow/WorkflowFactoryTest.java (hunk 2 above)
Comment (tsurdilo) [reply jbpm/jbpm-serverless-workflow/src/test/java/org/jbpm/serverless/workflow/WorkflowFactoryTest.java]: ok. updated

**Comment by tsurdilo** [comment]
@cristianonicolai It's a little different, with serverless workflow you can define functions which execute services. A function can define a type, and we use that to map to custom impls. 
So for example a "script" type maps to a script task, "service' type maps to workitems, and "decision" type maps to a human decision. Services like dmn , other workflows, other bpmn2 process and rules all fall under "services" which can be invoked either via things like rest workitem or a custom service invocation, so they would all fall under the "service" mapping which is already supported. Hope this helps.

**Comment by cristianonicolai** [comment]
@tsurdilo thanks for the details, I'm not following the spec in details so just thought that a decision could be either automated or done by a person.

**Comment by tsurdilo** [comment]
@cristianonicolai I added also KOGITO-2068 to this pr as there are no reviews and it was the next thing I was going to do anyways, and it affecting the same code as the original pr.

Also fixed issue with rule nodes being placed inside composite context nodes. this is tested by the serverless workflow tests in kogito-codegen.

**Comment by tsurdilo** [comment]
@MarianMacik @tiagodolphine Hi guys, would you mind taking a look when you have a chance please ?

File: jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/UserTaskModelMetaData.java [hunk 3]
```diff
@@ -160,7 +162,10 @@ private CompilationUnit compilationUnitInput() {
             Variable variable = variableScope.findVariable(entry.getValue());
 
             if (variable == null) {
-                throw new IllegalStateException("Task " + humanTaskNode.getName() +" (input) " + entry.getKey() + " reference not existing variable " + entry.getValue());
+                variable = processVariableScope.findVariable(entry.getValue());
```
Comment (tiagodolphine) [inline jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/UserTaskModelMetaData.java:165]: what about using 
Variable variable = Optional.ofNullable(variableScope.findVariable(entry.getValue()))
                    .orElse(processVariableScope.findVariable(entry.getValue()));

File: jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/UserTaskModelMetaData.java [hunk 4]
```diff
@@ -251,15 +256,18 @@ private CompilationUnit compilationUnitOutput() {
...
-                // check if given mapping is an expression
-                Matcher matcher = PatternConstants.PARAMETER_MATCHER.matcher(entry.getValue());
-                if (matcher.find()) {                    
-                    Map<String, String> dataOutputs = (Map<String, String>) humanTaskNode.getMetaData("DataOutputs");
-                    variable = new Variable();
-                    variable.setName(entry.getKey());
-                    variable.setType(new ObjectDataType(dataOutputs.get(entry.getKey())));
-                } else {
-                    throw new IllegalStateException("Task " + humanTaskNode.getName() +" (output) " + entry.getKey() + " reference not existing variable " + entry.getValue());
+                variable = processVariableScope.findVariable(entry.getValue());
```
Comment (tiagodolphine) [inline jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/UserTaskModelMetaData.java:259]: what about using
Variable variable = Optional.ofNullable(variableScope.findVariable(entry.getValue()))
.orElse(processVariableScope.findVariable(entry.getValue()));

Comment (tsurdilo) [reply jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/UserTaskModelMetaData.java]: updated

File: jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/UserTaskModelMetaData.java (hunk 3 above)
Comment (tsurdilo) [reply jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/UserTaskModelMetaData.java]: updated

**Comment by tsurdilo** [comment]
@tiagodolphine thanks fore review! Updated per your comments.

**Comment by tsurdilo** [comment]
@cristianonicolai can we just merge this? idk what the qe holdup might be about....

**Pull request merged**